remove exclusion of TS 2 options, add lib support

nicknisi opened the pull request:
* Remove restriction on what compilerOptions can be set
* remove filtering of TS 2.2 compilerOptions
* add support for 'list' option types
* Switch to processed compilerOptions
* add support for `compilerOptions.lib`

This PR removes the restrictions on setting options that aren't explicitly setup as options to allow new compilerOptions to pass through TypeDoc to TypeScript. It also sets up support for 'list' parameter types, and switches to use the processed compilerOptions, rather than `raw` so that the `lib` properties are properly parsed.

This should fix #315 and I believe #311.

blakeembrey commented on src/lib/utils/options/declaration.ts:
```diff
@@ -95,12 +95,13 @@ export class OptionDeclaration {
                     } else if (key in map) {
                         value = map[key];
-                    } else if (errorCallback) {
+                    } else if (!~values.indexOf(value) && errorCallback) {
```
Can we use the non-clever format of `indexOf`, cheers! 😄

nicknisi replied:
Removed! Sorry about that!

blakeembrey commented on src/lib/utils/options/options.ts:
```diff
@@ -152,10 +152,10 @@ export class Options extends ChildableComponent<Application, OptionsComponent> {
                 this.setValues(value, prefix + key + '.', errorCallback);
             } else {
-                this.setValue(prefix + key, value, errorCallback);
+                this.setValue(prefix + key, value);
```
Was there a reason behind this signature change?

blakeembrey commented on src/lib/utils/options/readers/tsconfig.ts:
```diff
@@ -51,31 +51,30 @@ export class TSConfigReader extends OptionsComponent {
 
-        let data = ts.readConfigFile(fileName, ts.sys.readFile).config;
-        if (data === undefined) {
+        let { config } = ts.readConfigFile(fileName, ts.sys.readFile);
```
Could we use `const` if there's no other mutation?

blakeembrey commented:
Sorry for the delay, I just had a couple of last comments. Once those are reviewed, I'm happy to merge!

nicknisi commented:
@blakeembrey Thanks! I rebased off the latest master and pushed up two commits to address your comments. Regarding the signature change, I actually modified my approach now that `OptionDeclaration#map` is a `Map`. The big thing I changed with this code (and the previous signature change) was that I am trying to let unknown-to-TypeDoc options simply pass through without throwing an error. That way, TypeDoc doesn't need to maintain a whitelist of `compilerOptions`, and should be more future proof.

Also, I checked in a number of changes to the built files (in a separate commit). These changes aren't related to this work, and appear when running the tests, even on master. I can remove this commit if you like.

blakeembrey commented:
Thanks @nicknisi! ❤️ 

I'm going to make a minor edit since it looks like you accidentally inserted a tab where there's spaces, then I'll merge 👍

blakeembrey commented:
Weird. I just pulled this in and tested, but I end up reverting all the `.html` changes you recently made. Do you have something locally affecting that?

nicknisi commented:
Yep, you're totally right... I didn't realize I had the themes repo `npm link`'d in!
